### ISSUE TITLE
Support JSON streaming formats besides application/stream+json

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-framework#21283]
**[Sebastiaan van Erk](https://jira.spring.io/secure/ViewProfile.jspa?name=sebster)** opened **[SPR-16742](https://jira.spring.io/browse/SPR-16742?redirect=false)** and commented

The mime type used by Spring Web Reactive for streaming json does not seem to be correct. The mime type "application/stream+json" is for an obsolete standard for activity streams:

https://tools.ietf.org/id/draft-snell-activity-streams-type-01.html

It has been superseded by the standard:

https://www.w3.org/TR/activitystreams-core/

with mime type "application/activity+json".

However, an activity stream is not the same as streaming json, so neither of these mime types should be used.

I was unable to find any standard mime type for json streaming, so I'm not sure what the resolution should be. The mime type should not be in the standards tree (since it is not a standard mime type), and the "x-" tree seems to be deprecated (since 1996) (see https://en.wikipedia.org/wiki/Media_type).

But I'm not sure that using a deprecated media type in the standards tree which is not officially registered, was invented for a different purpose, and has been superseded, is the way to go.

---

**Affects:** 5.0.5

**Issue Links:**
- #20331 Support Protobuf serialization in WebFlux

1 votes, 6 watchers

**Comment by membersound** [comment]
Is there a decision if `application/json+stream` will be deprecated in future? If so, what stream mediaType should I use if I'd want to output json as stream?

```
{"id":5, "name":"john"}
{"id":6, "name":"doe"}
{"id":7, "name":"jane"}
...
```

**Comment by rstoyanchev** [comment]
@sdeleuze maybe we should take another look for 5.2? 

I see [JSON Lines](http://jsonlines.org/) and [ndjson](http://jsonlines.org/ https://github.com/ndjson/ndjson-spec). The latter uses `"application/x-ndjson"` and is [used in Elasticsearch](https://www.elastic.co/guide/en/elasticsearch/reference/5.5/docs-bulk.html). There is also [JSON Sequences RFC](https://tools.ietf.org/html/rfc7464).

**Comment by makdeniss** [comment]
What is the status on this? FluentD f.e. defaults to ndjson now...

**Comment by raman-nbg** [comment]
Is there any update on this? What would be the best way to do JSON streaming right now?

**Comment by rstoyanchev** [comment]
Scheduling tentatively for 5.3 M1 to try and add support for additional JSON streaming formats. We could also deprecate "application/stream+json".

**Comment by rstoyanchev** [comment]
I've [deprecated "application/stream+json"](https://github.com/spring-projects/spring-framework/blob/683cc2eb7f7ce19720cac33e57d72277396bbc19/spring-web/src/main/java/org/springframework/http/MediaType.java#L231-L249) and [added "application/x-ndjson"](https://github.com/spring-projects/spring-framework/blob/683cc2eb7f7ce19720cac33e57d72277396bbc19/spring-web/src/main/java/org/springframework/http/MediaType.java#L219-L229). The latter is now also supported out of the box in the Jackson Encoder and Decoder. While it is not an officially registered media type, at least it's a common convention that seems to be in use while JSON lines doesn't have that. 

There is also JSON Text Sequences which can be used still with newline as the separator as long "application/json-seq" is registered with the Jackson Encoder and Decoder. As for other separators, I'm not sure that Jackson's non-blocking parser supports that.

In summary, use line-delimited JSON with "application/x-ndjson" supported as a media type out of the box and used as a replacement for "application/stream+json". Alternatively, register any other mime type and as long as it's a line-delimited JSON format it'll work the same. The mime type itself is just a convention for describing the format and as long as both sides understand it, it's enough.
